stage_9 create telegram bot buttons

Wonderfull23 opened the pull request.

Wonderfull23 committed:
stage_9 create telegram bt buttons

ievetrov commented on src/main/kotlin/additional/Telegram.kt:
```diff
@@ -1,29 +1,70 @@
     val botToken = args[0]
-    var updateId = 0
-    val bot = TelegramBotService()
+    var updateId = "0"
```
тут можно не делать его строкой,чтобы на 27 не преобразовывать туда сюда

ievetrov commented on src/main/kotlin/additional/TelegramBotService.kt:
```diff
@@ -2,23 +2,107 @@ package additional
 import java.net.http.HttpResponse
+import java.nio.charset.StandardCharsets
+
+const val statistics = "statistics_clicked"
```
константы нужно в UPPERCASE, проще глазами находить и отличать от переменных

ievetrov commented on src/main/kotlin/additional/TelegramBotService.kt:
```diff
@@ -2,23 +2,107 @@ package additional
+    fun sendQuestion(botToken: String, chatId: String?, question: Question?): String? {
+        val urlSendMessage = "https://api.telegram.org/bot$botToken/sendMessage"
+        var answers = ""
+        question?.variants?.forEachIndexed { index, word ->
```
норм, можно еще вместо forEachIndexed использовать mapIndexed. А объединять список получившихся строк при помощи .joinToString(separator = ","), тогда не нужно чесать что элемент последний

ievetrov commented on src/main/kotlin/additional/Telegram.kt:
```diff
@@ -1,29 +1,70 @@
+        } else if (data == learnWords) {
+            trainer.lastQuestion = checkNextQuestionAndSend(trainer, botToken, chatId)
+        }
+        else if (data != null && startsWith(data, CALLBACK_DATA_ANSWER_PREFIX)) {
```
А для чего метод startsWith? Можно ж применять прям экстеншн:
data?.startsWith(CALLBACK_DATA_ANSWER_PREFIX) == true

ievetrov replied:
Ну и в целом, видишь != null, попробуй заменить на null safe вызов

ievetrov commented on src/main/kotlin/additional/Telegram.kt:
```diff
@@ -1,29 +1,70 @@
-        if (text == "Hello") bot.sendMessage(botToken, chatId, text)
+        val data = dataRegex.find(updates)?.groups?.get(1)?.value
+        if (text?.lowercase() == "/start") bot.sendMenu(botToken, chatId)
+        if (data?.lowercase() == statistics) {
```
этот lowercase() не нужен, лишнее преобразование в нижний регистр

ievetrov commented on src/main/kotlin/additional/Telegram.kt:
```diff
@@ -1,29 +1,70 @@
+                    chatId,
+                    "Не правильно: ${trainer.lastQuestion?.correctAnswer?.original} - ${trainer.lastQuestion?.correctAnswer?.translate}"
+                )
+                trainer.lastQuestion = checkNextQuestionAndSend(trainer, botToken, chatId)
```
Правильнее было бы поле lastQuestion менять внутри класса LearnWordsTrainer, в момент вызова getNextQuestion() потому что именно тогда и меняется это поле.
А в твоем варианте есть риск забыть изменить trainer.lastQuestion, что приведет к багу.
В идеале trainer.lastQuestion должен быть извне только на чтение, и редактироваться внутри LearnWordsTrainer

ievetrov commented on src/main/kotlin/additional/LearnWordsTrainer.kt:
```diff
@@ -1,12 +1,19 @@
     val percent: Int,
-)
+) {
+    override fun toString(): String = "Выучено $learned из $total слов | $percent%"
```
Это прикольно, но если захочешь потом добавить английскую версию бота, нужно выносить.
Для нашей задачи подходит, оставляй так

ievetrov commented on src/main/kotlin/additional/Telegram.kt:
```diff
@@ -1,29 +1,70 @@
+        else if (data != null && startsWith(data, CALLBACK_DATA_ANSWER_PREFIX)) {
+            val index = data.substringAfter(CALLBACK_DATA_ANSWER_PREFIX).toInt()
+
+            if (trainer.lastQuestion != null && trainer.checkAnswer(trainer.lastQuestion!!, index + 1))
```
!! - этот оператор не зря такой восклицательный, его нужно избегать использовать.
В данном случае:
if(trainer.lastQuestion?.let { trainer.checkAnswer(it, index + 1) } == true) { ... }

Wonderfull23 committed:
stage_9
- changed type of updateId to Int
- renamed constants
- used mapIndexed instead of forEachIndexed
- deleted lowercase
- replaced !!
